### ISSUE TITLE
GCC Build Fails on AArch64

### ISSUE THREAD

**Issue by rgmiller** [spack/spack#5397]
I've been testing Spack out on our Cray Envoy system and have hit a problem with the GCC build.

The end of the spack-build.out file looks like this:
```
cp /home/sw/arm1/.spack/var/spack/stage/gcc-7.2.0-wodkpzprkhbwcsupyybm6w6ywayku554/gcc-7.2.0/gcc/../fixincludes/README-fixinc include-fixed/README
chmod a+r include-fixed/README
echo timestamp > stmp-int-hdrs
/tmp/xmr/spack-stage/spack-stage-UotD4k/gcc-7.2.0/spack-build/./gcc/xgcc -B/tmp/xmr/spack-stage/spack-stage-UotD4k/gcc-7.2.0/spack-build/./gcc/ -nostdinc -x c /dev/null -S -o /dev/null -fself-test=/home/sw/arm1/.spack/var/spack/stage/gcc-7.2.0-wodkpzprkhbwcsupyybm6w6ywayku554/gcc-7.2.0/gcc/testsuite/selftests
/tmp/xmr/spack-stage/spack-stage-UotD4k/gcc-7.2.0/spack-build/./gcc/cc1: error while loading shared libraries: libisl.so.15: cannot open shared object file: No such file or directory
Makefile:1933: recipe for target 's-selftest' failed
make[3]: *** [s-selftest] Error 1
rm gfortran.pod gcc.pod
make[3]: Leaving directory '/tmp/xmr/spack-stage/spack-stage-UotD4k/gcc-7.2.0/spack-build/gcc'
Makefile:4585: recipe for target 'all-stage2-gcc' failed
make[2]: *** [all-stage2-gcc] Error 2
make[2]: Leaving directory '/tmp/xmr/spack-stage/spack-stage-UotD4k/gcc-7.2.0/spack-build'
Makefile:22412: recipe for target 'stage2-bubble' failed
make[1]: *** [stage2-bubble] Error 2
make[1]: Leaving directory '/tmp/xmr/spack-stage/spack-stage-UotD4k/gcc-7.2.0/spack-build'
Makefile:931: recipe for target 'all' failed
make: *** [all] Error 2
```

You can see that it can't find libisl.so.15.

The install page at the GCC [wiki](https://gcc.gnu.org/wiki/InstallingGCC)  says that configuring with the --with-mpc, --with-mpfr & --with-gmp is not recommended because you have to add those paths to LD_LIBRARY_PATH every time you use gcc.  It appears that the same applies to --with-isl, too, with the additional caveat that you need to set LD_LIBRARY_PATH before even calling make.

The wiki page offers two recommended options:
1. Just use the existing system versions of libisl (and the others) or
2. After unpacking the source code, execute the contrib/download_prerequisites script which will fetch the mpc, mpfr, gmp and isl source packages extract them into the gcc source tree.  Then they'll automatically be built as part of gcc.  It seems to me that this would also mean that we could remove the dependencies on mpc, mpfr, gmp & isl from the gcc package.

Interestingly, I tried loading the isl module manually (which set LD_LIBRARY_PATH properly) but the build still failed.  At the moment, I'm assuming that LD_LIBRARY_PATH got reset somewhere during the build, but I haven't verified that.

Note that the build worked just fine when I tried it out on my x86_64 Fedora workstation, but that was probably because a sufficiently recent version of libisl was already installed in /usr/lib64.

**Comment by robertdfrench** [comment]
Thanks for reporting this @rgmiller

**Comment by pramodskumbhar** [comment]
> Interestingly, I tried loading the isl module manually (which set LD_LIBRARY_PATH properly) but the build still failed. At the moment, I'm assuming that LD_LIBRARY_PATH got reset somewhere during the build, but I haven't verified that.

Yes, you can find info [here](http://spack.readthedocs.io/en/latest/config_yaml.html?highlight=--dirty#dirty).

If you want to preserve environment, you ca use `--dirty` flag i.e. `spack install --dirty package`.

**Comment by rgmiller** [comment]
OK.  I'll try a build with the isl module loaded and the --dirty flag set and see what happens.

**Comment by robertdfrench** [comment]
@rgmiller that's my fault -- we have `--dirty` set by default and I forgot to tell you

**Comment by rgmiller** [comment]
Looks like building with the --dirty flag works.  I'm still curious why the package explicitly set the paths to the mpc, mpfr, gmp and isl libraries rather than using the download_prerequisites script that the GCC developers provide.  Does anyone know?

**Comment by rgmiller** [comment]
I made some changes to the gcc/package.py file to remove the dependencies on the gmp, mpfr, mpc and isl packages.  Instead, after downloading and unpacking the gcc tarball, package.py will call the 'download_prerequisites' script which will download gmp and the others directly into the gcc source tree.  From what I can tell, this has several advantages:

1. This is the recommended way to build gcc (see [https://gcc.gnu.org/wiki/InstallingGCC](https://gcc.gnu.org/wiki/InstallingGCC)) 
2. It saves me from having to manually load the (spack-generated) isl module and then build gcc with the --dirty flag
3. It ensures that when building gcc, you are actually building against the proper versions of the dependencies, and not against whatever is loaded on the system (which is what happens if you don't manually load modules and set the --dirty flag).

The changes work on my aarch64 testbed as well my x86_64 workstation.  If anybody is interested, I can try to push the changes back upstream.

**Comment by morganiain** [comment]
I encountered a similar problem building GCC on SLES 12/amd64. This occurred for versions 5.4, 6.2, and 7.2. In my case, enabling the binutils variant solved the problem.

**Comment by daveh86** [comment]
@rgmiller would you be able to push your changes up to a branch? I'd love to see

**Comment by rgmiller** [comment]
@daveh86  I've committed the change to a local branch on my workstation, but how do I push that branch up to this repository?  I don't seem to have write access (which doesn't surprise me).

**Comment by pramodskumbhar** [comment]
One way is to fork Spack repo under your GitHub account, push branch to your fork and then create PR from there.

**Comment by rgmiller** [comment]
OK. Pull request #5949 has been created.

**Comment by rgmiller** [comment]
FWIW: This more of a SLES12 issue than an AArch64 issue.  SLES12 includes an isl library that's too old for recent versions of gcc.  The fundamental problem is that when Spack builds gcc, it will link against the system version of libisl (and some other libraries) rather than the spack-built versions.

**Comment by tgamblin** [comment]
> The fundamental problem is that when Spack builds gcc, it will link against the system version of libisl (and some other libraries) rather than the spack-built versions.

☝️ @rgmiller: it looks like Spack build is not passing RPATHs to the gcc build properly.  There is some info [here](https://gcc.gnu.org/ml/gcc/2008-09/msg00212.html) on what to tweak to get an `RPATH` in for isl.

**Comment by rgmiller** [comment]
> it looks like Spack build is not passing RPATHs to the gcc build properly. There is some info here on what to tweak to get an RPATH in for isl.

Actually, I think it's even more complicated than that.  

I did some more testing on both my Fedora 26 workstation and the SLES 12 system.  On the Fedora system, the build works fine and  the resulting binaries **do** have RPATH set and **do** link against the correct, spack-built libraries.    On the SLES system, the build fails and if you look at the error message (see my first post in this thread) you'll see that problem was with one of the intermediate stage binaries.  A quick check with ldd shows that that particular binary does not try to use any of the spack-built libraries.

In short, it looks like RPATH is set for the final output of the gcc build, but it is **not** set for the intermediate stage(s).  The reason it works on Fedora 26 and not SLES 12 is that the default libraries in F26 are new enough that the intermediate stage binaries can get away with linking against them;  on SLES 12, the default version of libisl is too old for this to work.

**Comment by tgamblin** [comment]
> In short, it looks like RPATH is set for the final output of the gcc build, but it is not set for the intermediate stage(s).

There are args for this in the package currently, but they're only used when we build with binutils.  @mplegendre: do you know why that is?

Seems like you could add the dependency RPATHs to the args for intermediate stages and fix the issue.

**Comment by rgmiller** [comment]
@tgamblin I think that's the issue.  If I add `'--with-boot-ldflags={0}'.format(self.rpath_args)` to the options list at the top of `configure_args()`, then the build works fine.  It would also explain why enabling the binutils variant fixed the problem for @morganiain.

**Comment by tgamblin** [comment]
@rgmiller: Thanks for trying it out -- that's awesome!  Do you want to submit a PR with a fix?

**Comment by rgmiller** [comment]
I can, but I'd prefer to wait until @mplegendre answered your question above about why the flag was only set for the binutils variant.  Also, while we're looking at that code snippet, why is `static_bootstrap_flags` handled the way it is?  (ie: appended to two other elements of the `binutils_options` list rather than simply being added to the list once).

**Comment by tgamblin** [comment]
I dug into the history a bit, I don't think we need to ask @mplegendre.

I added the `--with-*-ldflags` arguments as unconditional args to `configure` back in 3e37903, and it looks like they stayed there until @alalazo made them only present witih `+binutils` in f475ec4.

@alalazo @trws @mplegendre: is there any particular reason to enable these only for `binutils` builds?

I think the solution is just to hoist those out to be regular `configure` args again.

**Comment by tgamblin** [comment]
I believe you want `static_bootstrap_flags` to be appended to each of the intermediate stage's LDFLAGS because those ensure that `libstdc++` and `libgcc` are linked statically for the bootstrap phase, which prevents the system `libstdc++` and `libgcc` from interfering with those stages.  I think you could also handle that by `RPATH`ing whatever directory those live in during the build, but the flags are easier b/c there are options for them and we don't have to understand the `gcc` build structure.

**Comment by mplegendre** [comment]
I don't know why those flags are only set for binutils.

**Comment by trws** [comment]
I’m not sure I’m up to date on gcc’s build process, but I don’t 
know any reason for that.

**Comment by tgamblin** [comment]
@rgmiller: I think you can go ahead and submit a PR with those flags enabled for the whole build if it works for you.

**Comment by rgmiller** [comment]
OK, I'm working on it.  It looks like enabling those flags breaks the x86_64 builds, though.  I need to figure that out before I issue a pull request.

**Comment by tgamblin** [comment]
@rgmiller thanks!
